### PR TITLE
Fix various Clang-12.0 compiler warnings

### PR DESCRIPTION
**Pull request by MegaManSec**
Fixes -Wrange-loop-construct, -Wmismatched-tags, and -Wshadow.

### CODE REVIEW

File: src/acl/Asn.cc [hunk 1]
```diff
@@ -595,7 +595,7 @@ ACLDestinationASNStrategy::match (ACLData<MatchType> * &data, ACLFilledChecklist
     const ipcache_addrs *ia = ipcache_gethostbyname(checklist->request->url.host(), IP_LOOKUP_IF_MISS);
 
     if (ia) {
-        for (const auto ip: ia->goodAndBad()) {
+        for (const auto &ip: ia->goodAndBad()) {
```
Comment (rousskov) [inline src/acl/Asn.cc:598]: Could you please let me know what warning did you get for this line. I agree that we should use a reference here, if possible, but I am curious what the compiler was warning us about in this a similar places.

File: src/http/RegisteredHeadersHash.cci [hunk 2]
```diff
@@ -41,7 +41,7 @@
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
 #line 24 "RegisteredHeadersHash.gperf"
-struct HeaderTableRecord;
+class HeaderTableRecord;
```
Comment (rousskov) [inline src/http/RegisteredHeadersHash.cci:44]: We should not be hand-editing generated files. AFAICT, this needs to be fixed in `src/http/RegisteredHeadersHash.gperf` (which also has a "generated from..." header, but _that_ header is lying IIRC).

You will need a matching gperf version to properly regenerate everything. See `git grep gperf scripts/source-maintenance.sh` for a regeneration command.

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for working on fixing these problems. Please add clang version to the PR title and (unique) warnings/codes to the PR description.

BTW, our PR title/description (i.e. the future commit message) format requirements are documented [elsewhere](https://github.com/measurement-factory/anubis/blob/master/README.md#commit-message). Following them should remove the M-failed-description label added by Anubis.

File: src/acl/Asn.cc (hunk 1 above)
Comment (rousskov) [reply src/acl/Asn.cc:598]: Never mind. I found the `-Wrange-loop-construct` answer in your branch commit and am resolving this change request. Please note that we normally _squash_ all branch commits, so, if there is useful information in them, we have to copy it to the PR description. This also helps avoid fishing for info in branch commits.

File: src/http/RegisteredHeadersHash.cci (hunk 2 above)
Comment (MegaManSec) [reply src/http/RegisteredHeadersHash.cci:44]: Ha! I originally fixed it in the gperf file, noticed it said "generated from", then moved it.

File: src/http/RegisteredHeadersHash.cci [hunk 3]
```diff
@@ -1,35 +1,35 @@
 /* C++ code produced by gperf version 3.1 */
-/* Command-line: gperf -m 100000 RegisteredHeadersHash.gperf  */
+/* Command-line: gperf -m 100000 ./RegisteredHeadersHash.gperf  */
```
Comment (MegaManSec) [inline src/http/RegisteredHeadersHash.cci:2]: I don't know why it has generated like this.
I ran `make -C src/http gperf-files` as suggested.

Comment (rousskov) [reply src/http/RegisteredHeadersHash.cci]: I fixed that and reformatted in 42991ff.

File: src/http/RegisteredHeadersHash.cci [hunk 4]
```diff
@@ -363,8 +363,8 @@ static const struct HeaderTableRecord HttpHeaderDefinitionsTable[] =
     {"Last-Modified", Http::HdrType::LAST_MODIFIED, Http::HdrFieldType::ftDate_1123, HdrKind::EntityHeader}
 };
 
-const struct HeaderTableRecord *
-HttpHeaderHashTable::lookup (const char *str, size_t len)
+const class HeaderTableRecord *
+    HttpHeaderHashTable::lookup (const char *str, size_t len)
```
Comment (rousskov) [inline src/http/RegisteredHeadersHash.cci:367]: I do not know why my astyle v2.04 indents this line while the official (probably) does not.

**Review by rousskov — APPROVED** [review]

File: src/http/RegisteredHeadersHash.cci (hunk 3 above)
Comment (MegaManSec) [reply src/http/RegisteredHeadersHash.cci]: Thank you.

**Comment by MegaManSec** [comment]
Btw, with the `S-waiting-for-committer` tag, if by 'committer' that means me (rather than someone to comment 'OK to test'), I don't know what i need to do.

**Comment by rousskov** [comment]
OK to test

**Comment by rousskov** [comment]
@MegaManSec, Semaphore CI does not see the latest commit on this branch. I suspect some GitHub notification got lost. It happens from time to time, and there is no Semaphore UI to _force_ a branch pull AFAICT. To wake Semaphore CI up, please feel free to push a dummy or trivial commit (perhaps removing the possibly bogus [indent](https://github.com/squid-cache/squid/pull/800#discussion_r606890647)?).

P.S. In this case, `S-waiting-for-committer` is my signal for other core developers (that have rights to add `M-clreared-for-merge`). I usually give others heads up for impending single-review PR mergers in hope to reduce human-human conflicts.

**Comment by MegaManSec** [comment]
Thank you for the explanation. I had originally thought `for-committer` meant something required on my side (as committer to the pull request), then took a look at some previous reports and understood it to mean "a privileged user needs to do something (as committer to the master branch).

**Comment by rousskov** [comment]
Your commit has [awaken](https://semaphoreci.com/squid/squid/branches/pull-request-800/builds/5) Semaphore CI, thank you!

P.S. We need to revamp and re-[publish](http://lists.squid-cache.org/pipermail/squid-dev/2017-August/009258.html) detailed explanations for these `S-...` labels and perhaps rename to avoid the overloaded term "committer" that means something different/special for Squid insiders than it is for new contributors.

**Review by yadij — APPROVED** [review]

**Pull request closed without merging**